server/server_setup.py

```diff
@@ -0,0 +1,241 @@
+
+    if "container" in container_data:
+        if "format" in container_data["container"]:
+            format_file = os.path.join(config_dir, container_data["container"]["format"] + ".yaml")
+            with open(format_file, "r") as ff:
+                format_data = yaml.load(ff, yaml.Loader)
+                return_data.update(format_data)
+
+    return return_data
+
+
+def create_server_config(
+    merlin_config_dir: str = MERLIN_CONFIG_DIR,
+    merlin_server_subdir: str = MERLIN_SERVER_SUBDIR,
+    merlin_server_config: str = MERLIN_SERVER_CONFIG,
+):
+    """
+    Create main configuration file for merlin server in the
+    merlin configuration directory. If a configuration already
+    exists it will not replace the current configuration and exit.
+
+    :param `merlin_config_dir`: location of main merlin configuration.
+    :param `merlin_server_subdir`: subdirectory for storing the server configuration.
+    :param `merlin_server_config`: server config file name
+    """
+    if not os.path.exists(merlin_config_dir):
+        LOG.error("Unable to find main merlin configuration directory at " + merlin_config_dir)
+        return False
+
+    config_dir = os.path.join(merlin_config_dir, merlin_server_subdir)
+    if not os.path.exists(config_dir):
+        LOG.info("Unable to find exisiting server configuration.")
+        LOG.info(f"Creating default configuration in {config_dir}")
+        try:
+            os.mkdir(merlin_config_dir + merlin_server_subdir)
+        except OSError as err:
+            LOG.error(err)
+            return False
+
+    files = [i + ".yaml" for i in CONTAINER_TYPES]
+    files.append(merlin_server_config)
+    for file in files:
+        file_path = os.path.join(config_dir, file)
+        if os.path.exists(file_path):
+            LOG.info(f"{file} already exists.")
+            continue
+        LOG.info(f"Copying file {file} to configuration directory.")
+        try:
+            shutil.copy(os.path.join(os.path.dirname(os.path.abspath(__file__)), file), config_dir)
+        except OSError:
+            LOG.error(f"Destination location {config_dir} is not writable.")
+            return False
+
+    print(pull_server_config())
+    return True
+
+
+def pull_server_image(server_dir: str = SERVER_DIR, image_name: str = IMAGE_NAME):
+    """
+    Fetch the server image using singularity.
+
+    :param `server_dir`: location of all server related files.
+    :param `image_name`: name of the image when fetched.
+    """
+    if not os.path.exists(server_dir):
+        LOG.info("Creating merlin server directory.")
+        os.mkdir(server_dir)
+
+    image_path = os.path.join(server_dir, image_name)
+
+    if os.path.exists(image_path):
+        LOG.info(f"{image_path} already exists.")
+        return False
+
+    LOG.info("Fetching redis image from docker://redis.")
+    subprocess.run(["singularity", "pull", image_path, "docker://redis"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

Comment:
These should be replaced with a fstring replaced version of the one from the yaml config.

Reply:
replaced with yaml config file